storybook/figmalinkssource.cpp

```diff
@@ -0,0 +1,66 @@
+#include "figmalinkssource.h"
+
+#include <QQmlEngine>
+
+#include "figmaio.h"
+#include "figmalinks.h"
+
+FigmaLinksSource::FigmaLinksSource(QObject *parent)
+    : QObject{parent}
+{
+    connect(&m_watcher, &QFileSystemWatcher::fileChanged,
+            this, [this](const QString &path) {
+        this->readFile();
+
+        if (!this->m_watcher.files().contains(path))
+            this->m_watcher.addPath(path);
+    });
+}
+
+void FigmaLinksSource::setFilePath(const QString& path)
+{
+    if (path == m_filePath)
+        return;
+
+    m_filePath = path;
+    readFile();
+    setupWatcher();
+    emit filePathChanged();
+}
+
+FigmaLinks* FigmaLinksSource::getFigmaLinks() const
+{
+    return m_figmaLinks;
+}
+
+void FigmaLinksSource::updateFigmaLinks(const QMap<QString, QStringList>& map)
+{
+    FigmaLinks *mapping = new FigmaLinks(map, this);
+
+    if (m_figmaLinks && qjsEngine(m_figmaLinks)) {
+        m_figmaLinks->setParent(nullptr);
+        QQmlEngine::setObjectOwnership(m_figmaLinks, QQmlEngine::JavaScriptOwnership);
+    }
+
+    m_figmaLinks = mapping;
+    emit figmaLinksChanged();
+}
+
+void FigmaLinksSource::readFile()
+{
+    QMap<QString, QStringList> figmaLinks = FigmaIO::read(m_filePath);
+    updateFigmaLinks(figmaLinks);
+}
+
+void FigmaLinksSource::setupWatcher()
+{
+    auto currentlyWatched = m_watcher.files();
+
+    if (currentlyWatched.size())
```

Comment:
```suggestion
    if (!currentlyWatched.isEmpty())
```

You don't really want to count the items

Reply:
For curiosity, see the difference in impl:

```cpp
    inline int size() const noexcept { return int(d->end - d->begin); }   // q6sizetype
    inline bool isEmpty() const noexcept { return d->end  == d->begin; }
```